### ISSUE TITLE
device graphs rotate on same device by device queue size / Alias argument/results

### ISSUE THREAD

**Issue by cjolivier01** [pytorch/xla#2194]
## ❓ Questions and Help

Hi, sorry to bother you, quick question:

When there's input and output aliased data handles to Execute (i.e. weights in and out), are the returned result data handles (in the result tuple) the same global data handles as the input argument's handle for which it aliases?  i.e.:

in: handles (23=features, 24=labels, **25**=weights, **26**=biases)

Do out handles tens to be:

* (27=loss, **25**=weight updates, **26**=bias updates)

or

* (27=loss, **_28_**=weight updates, **_29_**=bias updates)

Here I am referring tot he final weight update for the step. Additionally, for some reason that I don't fully understand, I tend to see two weight outputs which seem to be an intermediate calculation (I assume from what you're referring to in the build bias comment about A+B and such).

**Comment by cjolivier01** [comment]
There reason I asked was due to some unexpected behavior that turns out to be that the has differs and thus there is a separate compile for each item in the “device queue” of parallel loader. This could mean that the graph could change with every execution doesn’t it? I understand that the uid of each tensor differs since the tensors in the queue coexist and thus the hash of the tensors to sync differs and thus a different compile attempt (4 by default). 

Is this intended? i understand there is some chance that lower level caches may catch identical eventually-generated Hlo modules/ graphs after optimizations, but possibly not, I think.

For one device, is the intent to “rotate” possibly varying graphs for each sample?

Maybe I am misinterpreting something?

**Comment by dlibenzi** [comment]
I am not making sense of anything you are saying 😄 
In the parallel loader queues there are tensors, not IR graphs.
What does the parallel loader have to do with IR hashes, compilations, etc..?!?

**Comment by cjolivier01** [comment]
I sincerely apologize for this.  This was my bug.  I have a device with 400k+ cores and no memory bus and a batch's training time is, for all practical purposes, zero, such that everything else is the bottleneck.  In this case, for a simple fc network, what appears to be some sort of intermediate weight calculation are set as outputs (I assume it's some transient tensor created by autograd?), but aren't listed as aliases, so they become data that must be streamed off of the fabric, which slows everything down and frankly, nothing wants this data as an input off-fabric anyway that I have found, (not talking about the actual weight updates). So I was pruning these outputs and my bug was an accidental change which caused it to not prune them for every step after starting to proxy the execution over to the device (after the graph coalesces, since compile time is too long to do trivial setup computations on the device) which caused a new hash and thus a new graph.  I am sorry to have wasted your time.

**Comment by cjolivier01** [comment]
Actually, this turned out to be something different unrelated to my changes.  It's something with the dataset performing some sort of transformation that affects the graph on every pass, alternating back and forth between hashes, but only when device prefetch size > 1.  I reproduce by altering the test train mnist fake data to shape "batch_size, 784" instead of "batch_size, 1, 28, 28"
